### ISSUE TITLE
(typescript) $class variable is incorrectly highlighted as a keyword

### ISSUE THREAD

**Issue by aron-tw** [highlightjs/highlight.js#2516]
**Describe the issue**

**Which language seems to have the issue?**

**Are you using `highlight` or `highlightAuto`?** `highlight` 

...

**Sample Code to Reproduce**

https://jsfiddle.net/fsLxdrpm/1/
```ts
const ElementPrototype = Element.prototype;
const isBool = (v: any) => typeof v === 'boolean';
const $class = (v: any) => typeof v === 'boolean';
const whether = isBool(true);
const test = $class(true);
```

**Expected behavior**

1. `$class` prefix with `$` went wrong
2.  method with `class` name went wrong
3. method isn't be highlighted
4. Element isn't be highlighted 
5. isBool isn't be highlighted

**Comment by yyyc514** [comment]
> $class prefix with $ went wrong

This is a limitation of how we process keywords because technically the tiny area between `$` and `class` counts as a word boundary. Is this type of naming common in Typescript?

To fix it we'd have to hardcode a rule to match all `$[keyword]` instances and ignore them.

> method with class name went wrong

Ditto.

> method isn't be highlighted

What method? Also see my comment on isBool.

> Element isn't be highlighted

We don't have a completely list of types in our grammar... if you have a suggestion of which ones we should add we could discuss it and see if we can find an agreement - but I'm not sure we should "add them all" because there are a ton of them.

So we'd need some way to decide on the "major" ones...

> isBool isn't be highlighted

It wouldn't be, we don't highlight method invocation.

**Comment by yyyc514** [comment]
@egor-rogov Actually I wonder if we just extended `lexemes` to add `$` to it (even though it's not used in any keywords)... so now when looking for keywords it matches WHOLE identifiers (perhaps that's the best behavior anyways)... so it would see $class instead of class and go "nope, not a keyword".

See any downsides?

**Comment by yyyc514** [comment]
@allejo

**Comment by egor-rogov** [comment]
> @egor-rogov Actually I wonder if we just extended `lexemes` to add `$` to it (even though it's not used in any keywords)... so now when looking for keywords it matches WHOLE identifiers (perhaps that's the best behavior anyways)... so it would see $class instead of class and go "nope, not a keyword".

Yeah, I agree.
My first thought was about Perl (which uses `$` for scalars and has a separate rule for this), but no. In Perl `$` has its own meaning, while here it's just a part of identifier (right?).

**Comment by yyyc514** [comment]
Yes, in JS/TS $ and _ are just characters like any other. Hence jQuery $ and _ (underscore) libs.

**Comment by allejo** [comment]
> @egor-rogov Actually I wonder if we just extended `lexemes` to add `$` to it (even though it's not used in any keywords)... so now when looking for keywords it matches WHOLE identifiers (perhaps that's the best behavior anyways)... so it would see $class instead of class and go "nope, not a keyword".

I don't see any downsides to this since `$` isn't reserved. Just conventions in libraries like jQuery and Angular (they use `$$`).

**Comment by yyyc514** [comment]
Fix will go out in 10.1.
